File: Practice/17.letter-combinations-of-a-phone-number.java
import java.util.*;

public class Solution {

    private final String[] charmap = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };

    public List<String> letterCombinations(String digits) {
        LinkedList<String> res = new LinkedList<String>();
        if (digits.length() == 0 || digits == null) {
            return res;
        }
        res.add("");
        for (int i = 0; i < digits.length(); i++) {
            String string = charmap[digits.charAt(i) - '0'];
            while (res.peek().length() == i) {
                String s = res.remove();
                for (char c : string.toCharArray()) {
                    res.add(s + c);
                }

            }
        }
        return res;
    }
}

File: 17.letter-combinations-of-a-phone-number.js
/*
 * @lc app=leetcode id=17 lang=javascript
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
/**
 * @param {string} digits
 * @return {string[]}
 */

var letterCombinations = function (digits) {
  // 数字到号码的映射
  let map = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
  let result = [];
  let path = "";
  if (digits == null || digits.length == 0) return result;

  let backtrack = function (index) {
    if (path.length == digits.length) {
      result.push(path.slice());
      return;
    }
    let val = map[digits[index] - "2"];

    for (let char of val.split("")) {
      path += char;
      backtrack(index + 1);
      path = path.substring(0, path.length - 1);
    }
  };

  // 0 代表的是从digits的index=0开始
  backtrack(0);
  return res;
};
// @lc code=end
